interfaces/mount/backend.go

```diff
@@ -79,7 +80,19 @@ func (b *Backend) Setup(snapInfo *snap.Info, confinement interfaces.ConfinementO
 		return fmt.Errorf("cannot synchronize mount configuration files for snap %q: %s", snapName, err)
 	}
 	if err := UpdateSnapNamespace(snapName); err != nil {
-		return fmt.Errorf("cannot update mount namespace of snap %q: %s", snapName, err)
+		// try to discard the mount namespace and retry, but only if there aren't daemons in the snap
+		for _, app := range snapInfo.Apps {
+			if app.Daemon != "" && app.RefreshMode == "endure" {
+				return fmt.Errorf("cannot update mount namespace of snap %q when trying to update it, and cannot discard it and retry because it contains a daemon: %s", snapName, err)
+			}
+		}
+		logger.Debugf("cannot update mount namespace of snap %q; discarding namespace and retrying.", snapName)
+		if err = DiscardSnapNamespace(snapName); err != nil {
+			return fmt.Errorf("cannot discard mount namespace of snap %q when trying to update it: %s", snapName, err)
+		}
+		if err = UpdateSnapNamespace(snapName); err != nil {
+			return fmt.Errorf("cannot update mount namespace of snap %q: %s", snapName, err)
+		}
```

Comment:
I forgot to mention this but if the namespace is discarded this is a no-op, so this bit could be removed